### PR TITLE
1. CfeVesting: Vesting Account split and vasting account trace refactoring. 2. Genesis vesting pool split

### PR DESCRIPTION
**Pull request by pawel-borecki**
Vesting Account split

### CODE REVIEW

File: x/cfevesting/keeper/vesting_account_split.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+	for _, coin := range amountToUnlock {
+		if coin.Amount.GT(sdk.ZeroInt()) {
+			orignalVesting := orignalVestings.AmountOf(coin.Denom)
+			vestingCoin := vestingCoins.AmountOf(coin.Denom)
+			originalVestingDiffDec := sdk.NewDecFromInt(coin.Amount).Mul(sdk.NewDecFromInt(orignalVesting)).Quo(sdk.NewDecFromInt(vestingCoin))
+			originalVestingDiff := originalVestingDiffDec.TruncateInt()
+			vestingAcc.OriginalVesting = vestingAcc.OriginalVesting.Sub(sdk.NewCoin(coin.Denom, originalVestingDiff))
+			if !vestingCoin.Equal(coin.Amount.Add(vestingAcc.GetVestingCoins(ctx.BlockTime()).AmountOf(coin.Denom))) {
+				vestingAcc.OriginalVesting = vestingAcc.OriginalVesting.Sub(sdk.NewCoin(coin.Denom, sdk.NewInt(1))) // TODO is this enough ??
```
Comment (pablokulpa) [inline x/cfevesting/keeper/vesting_account_split.go:48]: 1 left for transaction?

File: x/cfevesting/keeper/vesting_account_split.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		return nil, sdkerrors.Wrapf(types.ErrNotExists, "account %s: not enough to unlock. locked: %s, to unlock: %s", ownerAddress, lockedCoins, amountToUnlock) // TODO some other error
+	}
+
+	vestingCoins := vestingAcc.GetVestingCoins(ctx.BlockTime())
+	orignalVestings := vestingAcc.OriginalVesting
+
+	for _, coin := range amountToUnlock {
+		if coin.Amount.GT(sdk.ZeroInt()) {
+			orignalVesting := orignalVestings.AmountOf(coin.Denom)
+			vestingCoin := vestingCoins.AmountOf(coin.Denom)
```
Comment (dawid-kruk) [inline x/cfevesting/keeper/vesting_account_split.go:43]: Rename to vestingAmount

File: x/cfevesting/keeper/msg_server_split_vesting.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	if len(amount) == 0 {
+		return sdkerrors.Wrapf(types.ErrParam, "split vesting coins - no coins to split %s", amount)
+	}
+
+	if amount.IsAnyNil() {
+		return sdkerrors.Wrapf(types.ErrParam, "split vesting coins - all coins of amount must not be nil: %s", amount)
+	}
+
+	if err := k.bank.IsSendEnabledCoins(ctx, amount...); err != nil {
+		return err
```
Comment (pablokulpa) [inline x/cfevesting/keeper/msg_server_split_vesting.go:49]: wrap this error

File: x/cfevesting/types/message_move_available_vesting.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	}
+	return []sdk.AccAddress{fromAddress}
+}
+
+func (msg *MsgMoveAvailableVesting) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(msg)
+	return sdk.MustSortJSON(bz)
+}
+
+func (msg *MsgMoveAvailableVesting) ValidateBasic() error {
```
Comment (dawid-kruk) [inline x/cfevesting/types/message_move_available_vesting.go:40]: Add additional validations in validate basic

File: x/cfevesting/types/message_move_available_vesting_by_denoms.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	_, err := sdk.AccAddressFromBech32(msg.FromAddress)
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid fromAddress address (%s)", err)
+	}
+	_, err = sdk.AccAddressFromBech32(msg.ToAddress)
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid toAddress address (%s)", err)
+	}
+	denomLen := len(msg.Denoms)
+	fmt.Printf("XXXXX: %d\n", denomLen)
```
Comment (dawid-kruk) [inline x/cfevesting/types/message_move_available_vesting_by_denoms.go:53]: delete print

File: x/cfevesting/keeper/vesting_account_split.go (hunk 1 above)
Comment (pawel-borecki) [reply x/cfevesting/keeper/vesting_account_split.go:48]: No, compensation for rounding/truncating in case of vesting amounts calculation

**Pull request merged**